### PR TITLE
Add support for EST Clients to advertise support for automated cert p…

### LINKED ISSUE #2: How do we know a node implements BCP-003-03?
The question is more to do with visibility of which nodes implement BCP-003-03 in an NMOS system. If you assume your entire system is BCP-003-03 but in fact you still have manually configured nodes which required certificates to be manually provisioned then you might end up in a situation where some endpoints fail because of expired certificates. Having some means of identifying which nodes implement BCP-003-03 seems beneficial as you would immediately also know which ones don't and require manual intervention.

### PR DESCRIPTION
**Pull request by JamesGibo**
…rovisioning, using the NMOS Node API's caps attribute

Close #2

### CODE REVIEW

File: docs/1.0. Certificate Provisioning.md [hunk 1]
```diff
@@ -206,6 +206,17 @@ An EST Client MUST provide a method to replace the manufacturer-issued TLS Clien
 
 An EST Client MAY ignore all timestamps in the certificate validity periods during the certificate provisioning process if it does not accurately know the current time, see [Bootstrapping Remote Secure Key Infrastructures - Section 2.6.1](https://tools.ietf.org/html/draft-ietf-anima-bootstrapping-keyinfra-41#section-2.6.1) for more information.
 
+If an EST Client implements the NMOS IS-04 Node API, then the EST Client SHOULD advertise that it supports BCP-003-03 using the 'caps' attribute of the Node API. Control interfaces can identify all Nodes which support BCP-003-03 by a key 'cert' and Uniform Resource Name (URN) having the base 'urn:x-nmos:control:cert' followed by a '/' character and the supported version.
```
Comment (andrewbonney) [inline docs/1.0. Certificate Provisioning.md]:
```suggestion
If an EST Client implements the NMOS IS-04 Node API, then the EST Client SHOULD advertise that it supports BCP-003-03 using the 'caps' attribute of the Node API. Control interfaces can identify all Nodes which support BCP-003-03 by a key 'cert' and Uniform Resource Name (URN) having the base 'urn:x-nmos:control:cert' followed by a '/' character and the supported version of this BCP.
```

We probably need a mechanism to cope with a client which supports multiple versions of the BCP. For Device 'controls' the array allows us to have multiple entries with different versioning.

**Review by andrewbonney — COMMENTED** [review]
This warrants discussion, but we may find it's 'safer' to define a tag for this purpose rather than using 'caps' given that the caps object is currently under discussion more generally. If sufficiently generic, the tag could be used in the same way to indicate support for other NMOS specs.

It should probably be specified whether this means the Node supports EST and it is currently active, or whether this may mean that EST is supported but its current status may be enabled or disabled.

**Comment by JamesGibo** [comment]
Adding Andrew's comment here so not lost from review
"We probably need a mechanism to cope with a client which supports multiple versions of the BCP. For Device 'controls' the array allows us to have multiple entries with different versioning."

File: docs/1.0. Certificate Provisioning.md [hunk 2]
```diff
@@ -206,13 +206,18 @@ An EST Client MUST provide a method to replace the manufacturer-issued TLS Clien
...
 
-**Example:** The 'caps' attribute of a NMOS Node
-```
+**Example:** The 'tags' attribute of a NMOS Node
+```json
 ...
-"caps": {
-  "cert": "urn:x-nmos:control:cert/v1.0",
+"tags": {
+  "cert": {
```
Comment (andrewbonney) [inline docs/1.0. Certificate Provisioning.md]:
```suggestion
  "urn:x-nmos:tag:cert": {
```

Matching the precedent in https://github.com/AMWA-TV/nmos-parameter-registers/blob/master/tags/grouphint.md a little more closely. The version could potentially make its way into the tag too.

**Review by prince-chrismc — COMMENTED** [review]
Since I missed the call, was there a motivation to using an object with an `enabled` key over just using the URN as a string?

**Comment by andrewbonney** [comment]
> Since I missed the call, was there a motivation to using an object with an `enabled` key over just using the URN as a string?

Were you thinking something like the following where true/false means enabled/disabled?
```
"urn:x-nmos:control:cert/v1.0": true
```

I realise given that we're looking at a tag this may actually have to be more like the following to match the IS-04 schemas, which is a little displeasing:
```
"urn:x-nmos:control:cert/v1.0": ["enabled"]
```

**Comment by prince-chrismc** [comment]
> Were you thinking something like the following where true/false means enabled/disabled?
> 
> ```
> "urn:x-nmos:control:cert/v1.0": true
> ```

Yes! That was exactly my thought
 
> I realise given that we're looking at a tag this may actually have to be more like the following to match the IS-04 schemas, which is a little displeasing:
> 
> ```
> "urn:x-nmos:control:cert/v1.0": ["enabled"]
> ```

Hmm, I agree that is the same as what we currently have in this PR.

**Review by prince-chrismc — APPROVED** [review]

File: docs/1.0. Certificate Provisioning.md [hunk 3]
```diff
@@ -206,6 +206,19 @@ An EST Client MUST provide a method to replace the manufacturer-issued TLS Clien
 
 An EST Client MAY ignore all timestamps in the certificate validity periods during the certificate provisioning process if it does not accurately know the current time, see [Bootstrapping Remote Secure Key Infrastructures - Section 2.6.1](https://tools.ietf.org/html/draft-ietf-anima-bootstrapping-keyinfra-41#section-2.6.1) for more information.
 
+If an EST Client implements the NMOS IS-04 Node API, then the EST Client SHOULD advertise that it supports BCP-003-03 using the 'tags' attribute of the Node API's `/self` endpoint. Control interfaces can identify all Nodes which support BCP-003-03 by a URN key 'urn:x-nmos:control:cert/', followed by the supported version of this BCP. If EST is currently active is signalled using the 'enabled' boolean.
```
Comment (garethsb) [inline docs/1.0. Certificate Provisioning.md]: Nitpick. "AMWA IS-04" or just "IS-04", not "NMOS IS-04".

**Comment by garethsb** [comment]
Agree with Andrew, that it needs to fit the `"tags"` schema, which is a bit awkward for this purpose:

```
...
"tags": {
  ...
  "urn:x-nmos:tag:certprov/v1.0": ["enabled"] // or "true"?
}
...
```

🤔

**Comment by JamesGibo** [comment]
I have updated the tag to match the tags schema and submitted a PR to nmos-parameter-registers to have the URN added.

https://github.com/AMWA-TV/nmos-parameter-registers/pull/20

File: docs/1.0. Certificate Provisioning.md [hunk 4]
```diff
@@ -206,6 +206,17 @@ An EST Client MUST provide a method to replace the manufacturer-issued TLS Clien
 
 An EST Client MAY ignore all timestamps in the certificate validity periods during the certificate provisioning process if it does not accurately know the current time, see [Bootstrapping Remote Secure Key Infrastructures - Section 2.6.1](https://tools.ietf.org/html/draft-ietf-anima-bootstrapping-keyinfra-41#section-2.6.1) for more information.
 
+If an EST Client implements the AMWA IS-04 Node API, then the EST Client SHOULD advertise that it supports BCP-003-03 using the 'tags' attribute of the Node API's `/self` endpoint. Control interfaces can identify all Nodes which support BCP-003-03 by a URN key 'urn:x-nmos:tag:certprov/', followed by the supported version of this BCP. When EST is enabled the value is set to `["enabled"]` and when disabled `["disabled"]`.
+
+**Example:** The 'tags' attribute of a NMOS Node
+```json
+...
+"tags": {
+  "urn:x-nmos:tag:certprov/v1.0": ["enabled"]
```
Comment (prince-chrismc) [inline docs/1.0. Certificate Provisioning.md]: :thought_balloon: Might this be reasonable? keep one tag that has a list of supported versions, similar to the api pages?

```suggestion
  "urn:x-nmos:tag:certprov": ["v1.0"]
```

Comment (garethsb) [reply docs/1.0. Certificate Provisioning.md]: I was wondering about this! The only downside I see is that didn't someone ask about distinguishing between support for the spec and that support being currently active/enabled? But I think it's entirely reasonable for the resource to indicate the current configuration, so personally I like this proposal better.

Comment (prince-chrismc) [reply docs/1.0. Certificate Provisioning.md]: > The only downside I see is that didn't someone ask about distinguishing between support for the spec and that support being currently active/enabled?

I would explicitly say listed versions are active, if the tag or any version is missing then it is unsupported/disabled. Another talking point for the next call?

Comment (garethsb) [reply docs/1.0. Certificate Provisioning.md]: Your proposal makes perfect sense to me.

Comment (andrewbonney) [reply docs/1.0. Certificate Provisioning.md]: Yes, that makes sense to me too. We just need to make sure the text is clear that this shouldn't be present if the EST support is currently disabled by configuration.

Comment (JamesGibo) [reply docs/1.0. Certificate Provisioning.md]: Sounds good, I have updated the docs to reflect this

File: docs/1.0. Certificate Provisioning.md [hunk 5]
```diff
@@ -206,6 +206,19 @@ An EST Client MUST provide a method to replace the manufacturer-issued TLS Clien
 
 An EST Client MAY ignore all timestamps in the certificate validity periods during the certificate provisioning process if it does not accurately know the current time, see [Bootstrapping Remote Secure Key Infrastructures - Section 2.6.1](https://tools.ietf.org/html/draft-ietf-anima-bootstrapping-keyinfra-41#section-2.6.1) for more information.
 
+If an EST Client implements the AMWA IS-04 Node API, then the EST Client SHOULD advertise that it supports BCP-003-03 using the 'tags' attribute of the Node API's `/self` endpoint. Control interfaces can identify all Nodes which support BCP-003-03 by a URN key 'urn:x-nmos:tag:certprov' and a list of enabled BCP-003-03 version. If BCP-003-03 is disabled the list MUST be empty. [NMOS Tags](https://github.com/AMWA-TV/nmos-parameter-registers/tree/master/tags)
```
Comment (andrewbonney) [inline docs/1.0. Certificate Provisioning.md]:
```suggestion
If an EST Client implements the AMWA IS-04 Node API, then the EST Client SHOULD advertise that it supports BCP-003-03 using the 'tags' attribute of the Node API's `/self` endpoint. Control interfaces can identify all Nodes which support BCP-003-03 by a URN key 'urn:x-nmos:tag:certprov' and a list of enabled BCP-003-03 versions. If BCP-003-03 is disabled the list MUST be empty. This tag is defined in the NMOS Parameter Registers under [NMOS Tags](https://github.com/AMWA-TV/nmos-parameter-registers/tree/master/tags).
```

**Comment by peterbrightwell** [comment]
Approved on call

**Pull request merged**